Implemented Metadata Definition UI

arunapa opened the pull request:
<img width="1624" alt="Screen Shot 2022-10-19 at 10 10 33 AM" src="https://user-images.githubusercontent.com/28983148/196731621-5c183e42-079b-447a-988e-b3eb7794db31.png">
<img width="1624" alt="Screen Shot 2022-10-19 at 10 11 37 AM" src="https://user-images.githubusercontent.com/28983148/196731628-4492bf78-360b-4d84-aa97-7bc2ccfbf463.png">

longshuicy commented:
I got a 422 error when posting definition:
<img width="524" alt="image" src="https://user-images.githubusercontent.com/13950475/196746991-5ec8b64f-d736-4b99-a172-ea1ff548112e.png">

Comparing your post body with the examples:
1. the context field should have another layer with the name of the definition as the key
2. you are missing the "name" of the definition, e.g. DOI at the root level
```
{
   "metadataName":"doi",
   "metadataDescription":"doi",
   "context":"https://schema.org/doi",
   "fields":[
      {
         "name":"doi",
         "list":false,
         "widgetType":"TextField",
         "config":{
            "type":"str",
            "options":""
         },
         "required":true
      }
   ]
}
```
vs
```
{
    "name" : "DOI",
    "description" : "Digital object identifier",
    "context" : {
        "doi" : "https://schema.org/DigitalDocument"
    },
    "fields" : [
        {
            "name" : "doi",
            "list" : false,
            "widgetType": "TextField",
            "config": {
                "type" : "str"
            },
            "required" : true
        }
    ]
}
```

longshuicy commented on frontend/src/components/Layout.tsx:
```diff
@@ -168,6 +168,17 @@ export default function PersistentDrawerLeft(props) {
+					<ListItem key={"newmetadata"} disablePadding>
+						<ListItemButton component={RouterLink} to="/new-metadata-definition">
+							<ListItemIcon>
+								<AddBox/>
```
I would change to another ICON just to differentiate with the new dataset icon. Maybe this: `import CreateIcon from '@material-ui/icons/Create';`?

arunapa replied:
Updated
<img width="1624" alt="Screen Shot 2022-10-27 at 1 19 01 AM" src="https://user-images.githubusercontent.com/28983148/198206283-1cd9ced2-28c0-4a93-a291-20199aaa1808.png">

longshuicy commented on frontend/src/components/datasets/Dataset.tsx:
```diff
@@ -276,6 +278,9 @@ export const Dataset = (): JSX.Element => {
 												<CloseIcon />
 											</IconButton>
+											<Button variant="contained" onClick={() => {setOpenPopup(true);}} sx={{ mt: 1, mr: 1, "alignItems": "right" }}>
+												Add new field
```
Add new metadata definition? Field is confusing to me.
Others, thoughts?

longshuicy commented on frontend/src/components/datasets/Dataset.tsx:
```diff
@@ -286,6 +291,22 @@ export const Dataset = (): JSX.Element => {
+												openPopup ?
+												<>
+													<Dialog open={openPopup} onClose={() => {setOpenPopup(false);}}>
+														<DialogTitle>Add a new field</DialogTitle>
```
Same here about the field

longshuicy commented on frontend/src/components/metadata/CreateMetadataDefinition.tsx:
```diff
@@ -0,0 +1,307 @@
+								onClick={clearForm}
+								sx={{ mt: 1, mr: 1, "alignItems": "right" }}>
+								Clear
+							</Button>
```
IMO it would be nice to add a "Back" at the very last step too. Sometimes I still want to change things until the very last...

arunapa replied:
I've updated the stepper code, now you can click directly on the stepper number to navigate back. I figured that a dding too many buttons would make the UI cluttered, hence adopted this approach instead :)

longshuicy commented on frontend/src/components/metadata/CreateMetadataDefinition.tsx:
```diff
@@ -0,0 +1,307 @@
+										placeholder="Please enter field data type (e.g. int, str, enum)"
+										name="Field Data Type"
+										value={input.config.type}
+										onChange={(event) => { handleInputChange(index, "type", event.target.value); }}
```
Since this right now is also an arbitrary list, could you also make it a dropdown?
```
"int": int,
"float": float,
"str": str,
"TextField": str,
"bool": bool,
"date": datetime.date,
"time": datetime.time,
"dict": dict,  # TODO: does this work?
"enum": str,  # TODO: need a way to validate enum,
"tuple": tuple,
```

longshuicy replied:
For both the widget type and data type drop downs, can we use more layman's term? e.g. TextField --> Text Box; Select --> Dropdown, Int --> Integer, Float --> Number
You are welcome to improvise a little bit :D

arunapa replied:
I've added human readable names in both drop downs

<img width="1624" alt="image" src="https://user-images.githubusercontent.com/28983148/198206993-defdd3e6-bddb-40d5-986a-9e6348fa34f6.png">

<img width="1624" alt="image" src="https://user-images.githubusercontent.com/28983148/198207039-1388ce5a-0ce6-4222-8e10-e38374187faf.png">

longshuicy commented on frontend/src/components/metadata/CreateMetadataDefinition.tsx:
```diff
@@ -0,0 +1,307 @@
+										onChange={(event) => { handleInputChange(index, "type", event.target.value); }}
+									/>
+									{(input.config.type == "enum") ? <>
+										<TextField
```
If you plan to address the multiple options of Enum in the next PR please create a task and add a TODO here :-)

arunapa replied:
Added To-dos wherever applicable

longshuicy commented on frontend/src/components/metadata/CreateMetadataDefinition.tsx:
```diff
@@ -0,0 +1,307 @@
+										</>
+										: 
+										<IconButton size="small" onClick={() => removeField(index)}>
+											<DeleteOutlineIcon />
```
When I click this delete Icon in this state, it behaves very weird
<img width="459" alt="image" src="https://user-images.githubusercontent.com/13950475/196750339-0cab5e2f-12e3-44c9-8559-f8b7b7919674.png">

arunapa replied:
Fixed the issue -- the stepper state wasn't being properly updated when an entry was deleted.

arunapa commented:
Some more improvements I've added in the latest PR:
1. Sanity checks while moving between stepper states (previously sanity was executed prior to invoking the API)
2. Added a new "preview" text box so that the user can know what metadata definition is being sent to the backend 
![image](https://user-images.githubusercontent.com/28983148/198207749-dd5eea8c-4fc2-49ce-b7fa-d65ea2ea407f.png)

3. Clicking on the stepper can move between states now

arunapa commented:
> 

I've addressed the following:
1. Making the capitalization consistent
2. Changing the default context example
3. Making the context example editable for the user to modify quickly while submitting the form
4. Checkboxes in same row

<img width="1624" alt="Screen Shot 2022-11-01 at 10 50 38 AM" src="https://user-images.githubusercontent.com/28983148/199277556-141b56e9-149b-405f-a0df-8f60701c9c3d.png">
<img width="1624" alt="Screen Shot 2022-11-01 at 10 50 44 AM" src="https://user-images.githubusercontent.com/28983148/199277559-cda9711e-ba12-41f0-8c7e-781d64172548.png">
